pkg/release/release.go

```diff
@@ -273,6 +279,25 @@ next:
 		}
 		logger.Log("info", "no changes", "phase", action)
 	case InstallAction:
+		// check if we need to migrate helm release from 2 to 3
+		if hr.GetHelmVersion(string(v1.HelmV2)) == string(v1.HelmV3) && hr.GetAnnotations()[Migrate2to3Annotation] == "true" {
+			clientV2, ok := r.helmClients.Load(string(v1.HelmV2))
+			if !ok {
+				status.SetStatusPhase(r.hrClient.HelmReleases(hr.Namespace), hr, v1.HelmReleasePhaseFailed)
+				errMessage := "could not init helm v2 client during 2to3 migration attempt"
+				logger.Log("error", errMessage, "phase", action)
+				errs = append(errs, fmt.Errorf(errMessage))
+				break
+			}
+
+			// run migration if helm2 release is found in tiller, otherwise, proceed with installation
+			_, err := clientV2.Get(hr.GetReleaseName(), helm.GetOptions{Namespace: hr.GetTargetNamespace()})
+			if err == nil {
+				action = Migrate2to3Action
+				goto next
+			}
+		}
```

Comment:
I would move this to `determineSyncAction`, and restructure it maybe a bit like this (pseudo code):

```go
if hr.GetAnnotations()[MigrateAnnotation] {
	switch hr.GetHelmVersion(v1.HelmV2) {
	case v1.HelmV3:
		...
	}
}
```